File: CodiceFiscale/model/identificativo.cpp
#include "identificativo.h"
#include <algorithm>


const std::string Identificativo::VOCALI("AEIOU");


Identificativo::Identificativo() : Sesso(), nome(""), cognome("") {}


Identificativo::Identificativo(std::string n, std::string c) : Sesso(), nome(n), cognome(c) {}


Identificativo::Identificativo(const Identificativo& a) : Sesso(a), nome(a.nome), cognome(a.cognome) {}


unsigned int Identificativo::contaConsonanti(std::string str) throw() {
    unsigned int c = 0; //c contiene il numero di consonanti trovate in str
    std::string::const_iterator it = str.begin();
    while(it < str.end()) {
        if(std::find(VOCALI.begin(), VOCALI.end(), *it) == VOCALI.end()) ++c;
        ++it;
    }
    return c;
}


std::string Identificativo::generaCodCognome() const throw() {
    //cognome contiene almeno 2 caratteri (da metodo controllo)
    std::string str; //conterra' codice cognome

    if(cognome.size() < 3) {
        //cognome ha meno di 3 caratteri, si aggiunge X alla fine

        //estrazione delle consonanti seguite dalle vocali, se presenti, nel loro ordine
        std::string::size_type pos = VOCALI.find(*(cognome.begin()));
        if(pos != std::string::npos) { //il primo carattere del cognome e' una vocale
            //controllo del secondo carattere
            pos = VOCALI.find(*(cognome.end()-1));
            if(pos == std::string::npos) { //il secondo carattere e' una consonante, str deve contenere prima la consonante, poi la vocale
                str.push_back(*(cognome.end()-1));
                str.push_back(*(cognome.begin()));
            } else str = cognome; //sono due vocali, str le contiene nel loro ordine
        } else str = cognome; //sono due consonanti, str le contiene nel loro ordine

        str.push_back('X'); //aggiunta X come terzo char
    } else {
        //cognome piu' lungo di 3 caratteri

        //costruzione cognome con le consonanti
        std::string::const_iterator it = cognome.begin();
        while(it < cognome.end() && str.size() < 3) {
            if(std::find(VOCALI.begin(), VOCALI.end(), *it) == VOCALI.end()) str.push_back(*it); // se e' una consonante viene inserita in str
            ++it;
        }

        //caso in cui le consonanti sono < 2
        it = cognome.begin();
        while(str.size() < 3){
            if(std::find(VOCALI.begin(), VOCALI.end(), *it) != VOCALI.end())
                str.push_back(*it);
            ++it;
        }
    }
    return str;
}


std::string Identificativo::generaCodNome() const throw() {
    //nome contiene almeno 2 caratteri (da metodo controllo)
    std::string str; //conterra' codice nome

    //controllo lunghezza nome
    if(nome.size() < 3) {
        //cognome ha meno di 3 caratteri, va aggiunta X alla fine

        //estrazione delle consonanti seguite dalle vocali, se presenti, nel loro ordine
        std::string::size_type pos = VOCALI.find(*(nome.begin()));
        if(pos != std::string::npos) { //il primo carattere del nome e' una vocale
            //controllo del secondo carattere
            pos = VOCALI.find(*(nome.end()-1));
            if(pos == std::string::npos) { //il secondo carattere e' una consonante, str deve contenere prima la consonante, poi la vocale
                str.push_back(*(nome.end()-1));
                str.push_back(*(nome.begin()));
            } else str = nome; //sono due VOCALI, str le contiene nel loro ordine
        } else str = nome; //sono due consonanti, str le contiene nel loro ordine

        str.push_back('X'); //aggiunta X come terzo char
    } else {
        //nome piu' lungo di 3 caratteri
        unsigned int cons = contaConsonanti(nome);
        unsigned int aux = 0;
        std::string::const_iterator it = nome.begin();

        //numero consonanti >3, si estraggono da nome la 1a, 3a e 4a
        while(it < nome.end() && str.size() < 3) {
            if(cons > 3) {
                if(std::find(VOCALI.begin(), VOCALI.end(), *it) == VOCALI.end()) {
                    if(aux != 1) str.push_back(*it);
                    ++aux;
                }
            } else {
                //numero consonanti <=3, si estraggono le consonanti di nome nel loro ordine
                if(std::find(VOCALI.begin(), VOCALI.end(), *it) == VOCALI.end()) str.push_back(*it);
            }
            ++it;
        }

        // caso in cui le consonanti sono < 2
        it = nome.begin();
        while(str.size() < 3){
            if(std::find(VOCALI.begin(), VOCALI.end(), *it) != VOCALI.end())
                str.push_back(*it);
            ++it;
        }
    }
    return str;
}

std::istream& operator>>(std::istream& is, Identificativo& i) throw(CFexception) {
    //possibili input contenenti spazi
    std::string nome, cognome;
    std::cout<<"Inserisci cognome: ";
    std::cin.ignore(); //scarta carattere newline nel buffer di input (dalla precedente digitazione della funzione)
    std::getline(is, cognome); //stop quando raggiunge carattere newline (newline letto e scartato)
    if(is.gcount() == 0) throw AntrCorto();
    std::cout<<"Inserisci nome: ";
    std::getline(is, nome);
    if(is.gcount() == 0) throw AntrCorto();
    nome.erase(remove_if(nome.begin(), nome.end(), isspace), nome.end()); //eliminazione spazi nome
    cognome.erase(remove_if(cognome.begin(), cognome.end(), isspace), cognome.end()); //eliminazione spazi cognome
    std::transform(nome.begin(), nome.end(), nome.begin(), ::toupper); //conversione minuscole in maiuscole
    std::transform(cognome.begin(), cognome.end(), cognome.begin(), ::toupper); //conversione minuscole in maiuscole
    i.nome = nome;
    i.cognome = cognome;
    return is;
}

void Identificativo::creaDaCodice(const std::string& cod) throw() {
    //cognome e nome conterranno il corrispondente codice estratto dal cf
    cognome = cod.substr(0,3);
    nome = cod.substr(3,3);
}


std::string Identificativo::generaCodice() const throw() {
    return generaCodCognome() + generaCodNome();
}


void Identificativo::controllo() throw(CFexception) {
    const std::string ALFABETO("ABCDEFGHIJKLMNOPQRSTUVWXYZ");

    //nome e cognome devono essere composti da almeno 2 caratteri
    if(nome.size() < 2 || cognome.size() < 2) throw AntrCorto();

    //controllo caratteri nome e cognome (le stringhe non contengono spazi)
    //nome e cognome possono contenere solo lettere dell'alfabeto
    std::string::size_type pos;
    for(std::string::const_iterator it = nome.begin(); it != nome.end(); ++it) {
        pos = ALFABETO.find(*it);
        if(pos == std::string::npos) throw ErrAntr();
    }
    for(std::string::const_iterator it = cognome.begin(); it != cognome.end(); ++it) {
        pos = ALFABETO.find(*it);
        if(pos == std::string::npos) throw ErrAntr();
    }
}


std::string Identificativo::toString() const throw() {
    return (cognome + " " + nome);
}


Identificativo::operator bool() const throw() {
    //controllo se oggetto identificativo ha valori assegnati (diversi da parametri di default)
    return cognome != "";
}

File: CodiceFiscaleJava/codicefiscalePkg/Identificativo.java
package codicefiscalePkg;


/**
 * La classe Identificativo rappresenta l'antroponimo
 * di una Persona, ed il sesso come sottooggetto.
 * Contiene funzioni utili per il controllo e
 * il calcolo del codice fiscale e del suo inverso.
 */
public class Identificativo extends Sesso {
	private String nome;
	private String cognome;
	
    private static final String VOCALI = ("AEIOU");

	
	public Identificativo() { this("",""); }

	
	public Identificativo(String n, String c) {
		super();
		nome = n;
		cognome = c;
	}


	//costruttore di copia
	public Identificativo(final Identificativo a) {
		super(a);
	    nome = a.nome;
	    cognome = a.cognome;
	}

	
	/**
	 * @return numero consonanti trovate in
	 * @param str
	 */
	private static int contaConsonanti(String str) {
		//str contiene vocali e consonanti
	    int conta_consonanti = 0;
	    for(int it = 0; it < str.length(); ++it) {
	    	char c = str.charAt(it); //estrazione carattere c da str in posizione it
	    	//ricerca c in VOCALI
	    	if(VOCALI.indexOf(c) == -1)
	    		//c non trovato in VOCALI => c e' consonante
	    		conta_consonanti +=1;
	    }
	    return conta_consonanti;
	}
	

	/**
	 * @return stringa di 3 caratteri che rappresentano
	 * cognome nel codice fiscale
	 */
	private String generaCodCognome() {
		//codice cognome sara' composto dalle consonanti, se presenti, seguite dalle vocali, nel loro ordine
		//cognome contiene almeno 2 caratteri
		
		String cod = ""; //conterra' codice cognome
		char c1, c2;
		
		//cognome contiene 2 caratteri, si aggiunge X alla fine
	    if(cognome.length() < 3) {
	    	c1 = cognome.charAt(0);
    		c2 = cognome.charAt(1);
    		if(VOCALI.indexOf(c1) != -1)
    			//il primo carattere di cognome e' una vocale
    			if(VOCALI.indexOf(c2) == -1)
    				//il secondo carattere di cognome e' una vocale
    				return c2 + c1 + "X"; //ritorna consonante + vocale + X
    		return cognome + "X"; //negli altri casi, ritorna cognome + X
	    }
	    
	    //cognome ha almeno 3 caratteri, si estraggono consonanti nel loro ordine
	    int it = 0;
	    while(it<cognome.length() && cod.length()<3) {
	    	c1 = cognome.charAt(it);
	    	if(VOCALI.indexOf(c1) == -1)
	    		//c1 e' una consonante
	    		cod = cod + c1;
	    	++it;
	    }
	    //se cod ha meno di 3 caratteri, si aggiungono le vocali di cognome nel loro ordine
	    if(cod.length() < 3) {
	    	it = 0;
	    	while(cod.length()!=3) {
	    		c1 = cognome.charAt(it);
	    		if(VOCALI.indexOf(c1) != -1)
	    			//c1 e' una vocale
	    			cod = cod + c1;
	    		++it;
	    	}
	    }
	    return cod; //cod e' composto da 3 caratteri
	}
	
	
	/**
	 * @return stringa di 3 caratteri che rappresentano
	 * nome nel codice fiscale
	 */
	private String generaCodNome() {
    	//codice nome sara' composto dalle consonanti, se presenti, seguite dalle vocali, nel loro ordine
		//nome contiene almeno 2 caratteri
		
		String cod = ""; //conterra' codice nome
		char c1, c2;
		
		//nome contiene 2 caratteri, si aggiunge X alla fine
	    if(nome.length() < 3) {
	    	c1 = nome.charAt(0);
    		c2 = nome.charAt(1);
    		if(VOCALI.indexOf(c1) != -1)
    			//il primo carattere di nome e' una vocale
    			if(VOCALI.indexOf(c2) == -1)
    				//il secondo carattere di nome e' una vocale
    				return c2 + c1 + "X"; //ritorna consonante + vocale + X
    		return nome + "X"; //negli altri casi, ritorna nome + X
	    }
	    
	    //nome ha almeno 3 caratteri, si estraggono consonanti nel loro ordine
        int n_cons = contaConsonanti(nome);
        int it = 0, conta = 0;
    	while(it<nome.length() && cod.length() < 3) {
        	c1 = nome.charAt(it);
    	    if(VOCALI.indexOf(c1) == -1 ) {
    	    	//c1 e' una consonante
    	    	if(n_cons >3) {
    	    		//n_cons > 3 => si prendono la 1a, 3a e 4a consonante del nome
	    	    	conta += 1;
	    	    	if(conta != 2)
	    	    		cod = cod + c1;
	    	    } else 
	    	    	//n_cons <= 3, si prendono le consonanti nell'ordine in cui si trovano
	    	    	cod = cod + c1;
    	    }
    	    ++it;
    	}

    	//se cod ha meno di 3 caratteri (quindi anche meno di 2 consonanti), si aggiungono le vocali di cognome nel loro ordine
	    if(cod.length() < 3) {
	    	it = 0;
	    	while(cod.length()!=3) {
	    		c1 = cognome.charAt(it);
	    		if(VOCALI.indexOf(c1) != -1)
	    			//c1 e' una vocale
	    			cod = cod + c1;
	    		++it;
	    	}
	    }
	    return cod; //cod e' composto da 3 caratteri
    }
    
    
	/** @return stringa del campo dati cognome. */
    public String estraiCognome() { return cognome; }
	
	
    /**
     * @param cod contiene 6 caratteri del codice fiscale
     * identificano nome e cognome.
     * I primi 3 caratteri vengono assegnati a cognome, gli ultimi 3 a nome.
     */
	public void creaDaCodice(final String cod) throws CFexception {
	    cognome = cod.substring(0,3);
	    nome = cod.substring(3, cod.length()-1);
	}
	
	
	/**
	 * @return stringa di 6 caratteri per il codice fiscale
	 * che rappresenta cognome e nome.
	 */
	public String generaCodice() {
	    return generaCodCognome() + generaCodNome();
	}
	

	/**
	 * @throws CFexception se nome o cognome hanno meno di 3 caratteri,
	 * oppure se nome o cognome contengono caratteri non in ALFABETO.
	 */
	public void controllo() throws CFexception {

	    //controllo lunghezza antroponimo
	    if(nome.length() < 2 || cognome.length() < 2) throw new CFexception("Nome o cognome troppo corti.");

	    //controllo caratteri antroponimo (nome e cognome non contengono spazi)
	    final String ALFABETO = ("ABCDEFGHIJKLMNOPQRSTUVWXYZ");
	    for(int it = 0; it < nome.length(); ++it) {
	    	char c = nome.charAt(it); //estrazione carattere c da nome in posizione it
	    	if(ALFABETO.indexOf(c) == -1) throw new CFexception("Nome non valido."); //controllo che c sia in ALFABETO
	    }
	    for(int it = 0; it != cognome.length(); ++it) {
	    	char c = cognome.charAt(it); //estrazione carattere c da cognome in posizione it
	    	if(ALFABETO.indexOf(c) == -1) throw new CFexception("Cogome non valido."); //controllo che c sia in ALFABETO
	    }
	}
	

	/**
	 * overrides java.lang.Object.toString
	 * @return stringa descrittiva dei campi dati di Identificativo
	 */
	public String toString() {
	    return (cognome + " " + nome);
	}
	
}
